csrc/abstract_tensor_schedule.cpp

```diff
@@ -0,0 +1,261 @@
+  //! scheduled ValGroup, otherwise we would not know where to place this
+  //! IterDomain.
+  void findNearestProducers() {
+    // Record all ValGroups that are producers of those in abstract_. This lets
+    // us find starting IterDomains for scheduling concrete's loop domain.
+    // TODO: this should probably use a new more general utility that resembles
+    // StmtSort::getExprsTo in val_graph_visitor.h
+    std::unordered_set<ValGroup> scheduled_val_groups;
+    std::stack<ValGroup> vg_stack;
+    for (const AbstractId& abs_id : abstract_.domain) {
+      vg_stack.push(abstractIdToValGroup(abs_id));
+    }
+    while (!vg_stack.empty()) {
+      ValGroup vg = vg_stack.top();
+      vg_stack.pop();
+      bool inserted = scheduled_val_groups.insert(vg).second;
+      if (!inserted) {
+        // Avoid cycles
+        continue;
+      }
+      for (const ExprGroup& eg : graph_->getDefinitions(vg)) {
+        for (Expr* e : *eg) {
+          for (Val* inp : e->inputs()) {
+            ValGroup vg_inp = graph_->toGroup(inp);
+            vg_stack.push(vg_inp);
+          }
+        }
+      }
+    }
+
+    // Now traverse c2p from concrete loop domain, stopping when we find a
+    // scheduled ValGroup
+    std::stack<IterDomain*> id_stack;
+    for (IterDomain* id : concrete_->getLoopDomain()) {
+      id_stack.push(id);
+    }
+    while (!id_stack.empty()) {
+      IterDomain* id = id_stack.top();
+      id_stack.pop();
+      ValGroup g = graph_->toGroup(id);
+      if (scheduled_val_groups.find(g) != scheduled_val_groups.end()) {
+        concrete_ids_.emplace(g, id);
+        continue;
+      }
+      NVF_ERROR(
+          id->definition() != nullptr,
+          "Root IterDomain ",
+          id->toString(),
+          " does not appear in the history of any ValGroups in abstract tensor");
+      for (Val* inp : id->definition()->inputs()) {
+        if (auto inp_id = dynamic_cast<IterDomain*>(inp)) {
+          id_stack.push(inp_id);
+        }
+      }
+    }
+  }
+
+  IterDomain* replayAbstractId(AbstractId abs_id) {
+    ValGroup g = abstractIdToValGroup(abs_id);
+
+    // This holds ValGroups that we cannot compute from concrete_'s root. When
+    // we detect that any inputs of an ExprGroup are in these groups, we know
+    // that we cannot compute that ExprGroup so we should skip it.
+    std::unordered_set<ValGroup> uncomputable_groups;
+
+    std::stack<ValGroup> vg_stack({g});
+    // Strategy: fill out concrete_ids_ in the direction of g
+    auto propagate = [&](ValGroup vg) {
+      // We need to try and produce an ID for this group, so we look at
+      // definitions for the Vals in this group and check whether we have their
+      // inputs computed yet (concrete_ids_) and whether they can be computed
+      // (uncomputable_groups).
+      std::vector<ValGroup> uncomputed_producer_groups;
+      for (const ExprGroup& eg : graph_->getDefinitions(vg)) {
+        for (Expr* e : *eg) {
```

Comment:
Why do we need this `for (Expr* e : *eg)`, instead of just use `graph_->inputGroups`?

Reply:
If we only needed to find all input `ValGroup`s then `inputGroups` would be better. However, we also need to know what kind of `ExprGroup` we have in order to create `id_expr`. To do that I check the type of some of the `Expr`s in the group until I find one all of whose inputs are computable using IDs from the given TV, then I go ahead and evaluate it.

Reply:
> However, we also need to know what kind of `ExprGroup`

Would `expr_group->front()` sufficient to know the expr type?

Reply:
I am not sure. I _think_ it will always be the case that every `Expr* e` in an `ExprGroup` will have the same set of `ValGroup`s if you look at `e->inputs()` in the permissive map, and all those `Expr`s should also have compatible (proven equal) attributes. I am not sure if that is a fundamental assumption about the graph, or if it is just coincidentally true for the limited set of IterDomain expressions we currently have.

It would be simpler to just handle `expr_group->front()` and we could later try this current approach if it's needed.

Reply:
We only map exprs that has matching type and attributes:
https://github.com/NVIDIA/Fuser/blob/65799ed56f66b671dfbd51336fa302400e2bdf45/csrc/val_graph.cpp#L338

Reply:
OK great. I'll change to only checking the first Expr in the group as you suggested.